components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
             .pop_front()
             .expect("Reader must have read request when remove is called into.")
     }
+
+    /// <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
+    pub fn empty_read_requests(&self) {
+        // step 2
+        self.read_requests.borrow_mut().clear();
+    }
 }
 
 impl ReadableStreamDefaultReaderMethods for ReadableStreamDefaultReader {
     /// <https://streams.spec.whatwg.org/#default-reader-read>
     fn Read(&self) -> Rc<Promise> {
         let promise = Promise::new(&self.reflector_.global());
 
-        self.stream
-            .perform_pull_steps(ReadRequest::Read(promise.clone()));
+        if let Some(stream) = self.stream.get() {
+            stream.perform_pull_steps(ReadRequest::Read(promise.clone()));
+        } else {
+            promise.reject_error(Error::Type("stream is undefined".to_owned()));
+        }
 
         promise
     }
 
     /// <https://streams.spec.whatwg.org/#default-reader-release-lock>
     fn ReleaseLock(&self) {
-        if self.stream.is_readable() {
-            self.closed_promise.reject_native(&());
+        // step 1
+        if self.stream.get().is_none() {
+            return;
+        }
+
+        // <https://streams.spec.whatwg.org/#readable-stream-reader-generic-release>
+        // step 2.1.2
+        assert!(self.stream.get().is_some());
+        // step 2.1.3
+        let stream = self.stream.get().unwrap();
+        assert!(stream.is_locked());
+
+        if stream.is_readable() {
+            // step 2.1.4
+            self.closed_promise
+                .reject_error(Error::Type("stream state is not readable".to_owned()));
+        } else {
+            // step 2.1.5
+            self.closed_promise.reject_error(Error::Type(
+                "Cannot release lock due to stream state.".to_owned(),
+            ));
         }
 
-        // TODO: https://streams.spec.whatwg.org/#readable-stream-reader-generic-release
+        // step 2.1.6
+        self.closed_promise.set_promise_is_handled();
+
+        // step 2.1.7
+        stream.perform_release_steps();
+
+        // step 2.1.8 & 2.1.9
+        self.stream.set(None);
+
+        // step  3.1
+        let error = Error::Type("No chunks are available because the stream is errored".to_owned());
 
-        // TODO: use TypeError.
         // <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
+        // step 3.2
+        self.empty_read_requests();
+        // step 3.3
         for request in self.read_requests.borrow_mut().drain(0..) {
-            request.error_steps();
+            request.error_steps(error.clone());
```

Comment:
I think we should clone `self.read_requests` somewhere before clearing it. So that on the step 3.3 we can perform `error_steps()` to those `read_requests`

Reply:
or maybe we can do `drain` only?

Reply:
indeed thanks